File: frontend/src/api/jiraConnection.ts
import { apiRequest } from "./client";
import type {
  JiraConnectionConfig,
  JiraConnectionPayload,
  JiraConnectionTestResult,
} from "../types/jiraConnection";

export const jiraConnectionApi = {
  get(): Promise<JiraConnectionConfig> {
    return apiRequest<JiraConnectionConfig>("/api/settings/jira");
  },

  save(payload: JiraConnectionPayload): Promise<JiraConnectionConfig> {
    return apiRequest<JiraConnectionConfig>("/api/settings/jira", {
      method: "PUT",
      body: payload,
    });
  },

  test(payload: JiraConnectionPayload): Promise<JiraConnectionTestResult> {
    return apiRequest<JiraConnectionTestResult>("/api/settings/jira/test", {
      method: "POST",
      body: payload,
    });
  },
};

File: backend/app/api/jira_connection.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request, status

from app.models.jira_connection import (
    JiraConnectionConfig,
    JiraConnectionPayload,
    JiraConnectionTestResult,
)
from app.services.jira_connection_service import JiraConnectionService

router = APIRouter(prefix="/api/settings/jira", tags=["jira-settings"])


def get_jira_connection_service(request: Request) -> JiraConnectionService:
    return request.app.state.jira_connection_service


@router.get("", response_model=JiraConnectionConfig)
async def get_jira_connection(
    service: Annotated[JiraConnectionService, Depends(get_jira_connection_service)],
) -> JiraConnectionConfig:
    return await service.get_connection()


@router.put("", response_model=JiraConnectionConfig)
async def save_jira_connection(
    payload: JiraConnectionPayload,
    service: Annotated[JiraConnectionService, Depends(get_jira_connection_service)],
) -> JiraConnectionConfig:
    try:
        return await service.save_connection(payload)
    except ValueError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(error),
        ) from error


@router.post("/test", response_model=JiraConnectionTestResult)
async def test_jira_connection(
    payload: JiraConnectionPayload,
    service: Annotated[JiraConnectionService, Depends(get_jira_connection_service)],
) -> JiraConnectionTestResult:
    try:
        return await service.test_connection(payload)
    except ValueError as error:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(error),
        ) from error
